libdnf/comps/environment/environment.cpp

```diff
@@ -0,0 +1,317 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/comps/environment/environment.hpp"
+
+#include "solv/pool.hpp"
+#include "utils/xml.hpp"
+
+#include "libdnf/base/base.hpp"
+
+extern "C" {
+#include <solv/dataiterator.h>
+#include <solv/knownid.h>
+#include <solv/pool.h>
+#include <solv/solvable.h>
+}
+
+#include <libxml/tree.h>
+
+#include <string>
+
+
+namespace libdnf::comps {
+
+
+Environment::Environment(libdnf::Base & base) : base(base.get_weak_ptr()) {}
+
+
+Environment & Environment::operator+=(const Environment & rhs) {
+    this->environment_ids.insert(this->environment_ids.begin(), rhs.environment_ids.begin(), rhs.environment_ids.end());
+    return *this;
+}
+
+
+// Search solvables that correspond to the environment_ids for given key
+// Return first non-empty string
+std::string lookup_str(libdnf::solv::Pool & pool, std::vector<EnvironmentId> environment_ids, Id key) {
+    for (EnvironmentId environment_id : environment_ids) {
+        auto value = pool.lookup_str(environment_id.id, key);
+        if (value) {
+            return value;
+        }
+    }
+    return "";
+}
+
+
+std::string Environment::get_environmentid() const {
+    libdnf::solv::Pool & pool = get_pool(base);
+
+    std::string solvable_name(lookup_str(pool, environment_ids, SOLVABLE_NAME));
+    if (solvable_name.find(":") == std::string::npos) {
+        return "";
+    }
+    return solvable_name.substr(solvable_name.find(":") + 1);
+}
+
+
+std::string Environment::get_name() const {
+    libdnf::solv::Pool & pool = get_pool(base);
+    return lookup_str(pool, environment_ids, SOLVABLE_SUMMARY);
+}
+
+
+std::string Environment::get_description() const {
+    libdnf::solv::Pool & pool = get_pool(base);
+    return lookup_str(pool, environment_ids, SOLVABLE_DESCRIPTION);
+}
+
+
+std::string Environment::get_translated_name(const char * lang) const {
+    libdnf::solv::Pool & pool = get_pool(base);
+
+    // Go through all environment solvables and return first translation found.
+    for (EnvironmentId environment_id : environment_ids) {
+        Solvable * solvable = pool.id2solvable(environment_id.id);
+        if (const char * translation = solvable_lookup_str_lang(solvable, SOLVABLE_SUMMARY, lang, 1)) {
+            // Return translation only if it's different from the untranslated string
+            // (solvable_lookup_str_lang returns the untranslated string if there is no translation).
+            const char * untranslated = solvable_lookup_str(solvable, SOLVABLE_SUMMARY);
+            if (translation != untranslated && strcmp(translation, untranslated) != 0) {
+                return std::string(translation);
+            }
+        }
+    }
```

Comment:
> There'd be an issue with sharing it with the goup code in that `GroupId` and `EnvironmentId` are different types, which can be solved by templating the function.

Oh, cool, I'll do that then.

> Another issue is if the translation isn't found (not handled in my example code above). I'm not sure why you actually compare the translation with the untranslated string, since at the end of each function you seem to return the untranslated string anyway? Granted it's retrieved a bit differently from libsolv, but don't they actually result in the same string being returned? IIRC I've once checked and libsolv's `pool_lookup_str()` and `solvable_lookup_str()` return the same thing, barring some special cases which I don't think apply here.

I compare the translation with the untranslated string, because libsolv returns the untranslated string if the translation doesn't exist. And I have to check every time, because theoretically there could be different untranslated string for every solvable. I could get rid of exactly one search (the one at the end) by remembering the first one, but I think it would be even more confusing.

> Then this and the `lookup_str()` function above could be shared in a file (possibly called `libdnf/comps/pool_utils.{hpp,cpp}` or similar). I recall we discussed sharing the `lookup_str()` right? That one small function maybe wasn't worth it but if there were two and this second one a bit more complex, seems sharing the code would be better).

Ok, thanks.